File: leetcode/math/2979_Most_Expensive_Item_That_Can_Not_Be_Bought/codes/dp.c
#include <stdio.h>
#include <stdbool.h>
#include <string.h>

int mostExpensiveItem(int primeOne, int primeTwo) {
    int max_val = primeOne * primeTwo;
    bool dp[max_val + 1];
    memset(dp, 0, sizeof(dp));
    dp[0] = true;

    for (int i = 1; i <= max_val; ++i) {
        if (i >= primeOne && dp[i - primeOne])
            dp[i] = true;
        if (i >= primeTwo && dp[i - primeTwo])
            dp[i] = true;
    }

    for (int i = max_val; i >= 0; --i) {
        if (!dp[i])
            return i;
    }

    return 0; 
}

File: leetcode/math/2979_Most_Expensive_Item_That_Can_Not_Be_Bought/codes/dp.py
class Solution:
    def mostExpensiveItem(self, primeOne: int, primeTwo: int) -> int:
        max_val = primeOne * primeTwo
        dp = [False] * (max_val + 1)
        dp[0] = True

        for i in range(1, max_val + 1):
            if i >= primeOne and dp[i - primeOne]:
                dp[i] = True
            if i >= primeTwo and dp[i - primeTwo]:
                dp[i] = True

        for i in range(max_val, -1, -1):
            if not dp[i]:
                return i
